Open markdown reference

sim590 opened the pull request:
I've added a functionnality for opening a markdown reference under the cursor. It has be inspired by @christoomey's [function](https://github.com/christoomey/dotfiles/blob/f3f9da69586503823de9ff075fbf627ec2bb6fb5/vim/rcfiles/prose#L90-L106). In order not to depend on any (I *hope* I'm not mistaking here...) platform, I've chosen to use `:normal gx` from Vim's `pi_netrw` plugin (which comes with **Vim 7.4**, I guess).

PLUS : I've fixed a minor typo inside the README file ! ;)

christoomey commented:
Thanks for taking this on! I love the idea of aligning with `gx`.

If possible, I'd be interested in trying to go a bit further and implement via `gx` directly, such that `gx` on a markdown `[link][]` would use the markdown behavior as you've defined it, otherwise fall back to `normal! gx` (might need to use the function it maps to, `call netrw#NetrwBrowseX(expand("<cfile>"),0)`).

I think you could use the following to rely on the syntax highlighting to determine if your cursor is on a markdown link: 

``` vim
function! s:OnMarkdownLink()
  return synIDattr(synID(line("."),col("."),1),"name") == "markdownLinkText"
endfunction
```

Then use as

``` vim
function! s:MarkdownAwareGX()
  if s:OnMarkdownLink()
    call s:OpenMarkdownLink()
  else
    call netrw#NetrwBrowseX(expand("<cfile>"),0
  endif
endfunction
command! MarkdownAwareGX call <sid>MarkdownAwareGX()
nmap <buffer> gx :MarkdownAwareGX<cr>
```

Haven't fully tested, but I think this should all work.

christoomey commented on plugin/quicklink.vim:
```diff
@@ -95,4 +95,30 @@ function! OpenLinkOnCurrentLine()
+  let @a = ""
+  normal "ayi]
+  let escaped_link_name = escape(getreg('a'), '&')
+  let pattern = '\v^\['.escaped_link_name.'\]: (%(ftp[s]?|http[s]?):\/\/\S+)>'
```
Can you rename this variable to be a bit more explicit? `link_target_pattern` or similar?

christoomey commented on plugin/quicklink.vim:
```diff
@@ -95,4 +95,30 @@ function! OpenLinkOnCurrentLine()
+  let pattern = '\v^\['.escaped_link_name.'\]: (%(ftp[s]?|http[s]?):\/\/\S+)>'
+  " go to link location
+  let found = search(pattern, 'e')
+  if found
```
If you rename the pattern variable, I think it makes sense to inline this call:

``` vim
if search(link_target_pattern, 'e')
...
```

sim590 commented:
Yes! I've thought about that and it was the way I wanted to go, but I couldn't find a way to call  `netrw#NetrwBrowseX(expand("<cfile>"),0)` directly, it would say that the function was undefined. I've just found how to do it. I'll change for that because it's much more elegant!

sim590 commented:
Works great !

christoomey commented:
@sim590 Thanks for working on this! This looks great, and I am really happy with being able to simplify down to the consistent `gx` key mapping for all link opening. I have a couple of comments which I'll add inline, but just wanted to say thanks for moving this along.

christoomey commented on README.md:
```diff
@@ -20,6 +20,19 @@ the search. A results window will pop up with the following keymaps:
+This functionnality uses `pi_netrw` (`:normal gx`) in order to
+open the link, so it will open it using the program specified by the global
+variable `g:netrw_browsex_viewer` (see `:h netrw-gx`). Also, `:normal gx` will
+behave normally if on normal link.
```
Since we have two distinct features here, can we break down the usage section to match. Something like:

``` md
### URL Searching

The main feature of `vim-quicklink` is to search for and insert a relevant
link without needing to leave vim. You can do this by...

...

### Link Opening

`vim-quicklink` also updates the `gx` mapping from `netrw` to work with markdown link format. The default
behavior of `gx` when your cursor is over a URL will still work and is unchanged, but you can now also use
`gx` on a markdown link such as `[webapi-vim][]` in this readme and `vim-quicklink` will find and open the
associated URL using `netrw`.
```

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+    return synIDattr(synID(line("."),col("."),1),"name") == "markdownLinkText" ||
+                \ synIDattr(synID(line("."),col("."),1),"name") == "markdownLinkTextDelimiter" ||
+                \ synIDattr(synID(line("."),col("."),1),"name") == "mkdLink" ||
+                \ synIDattr(synID(line("."),col("."),1),"name") == "mkdDelimiter"
```
I think you can simplify and clarify this by saving off the `synIDattr` return value and using a list count check:

``` vim
" Define this as a constant(ish) near the top of the script
let s:MARKDOWN_LINK_SYNTAX_IDS = [
  \ "markdownLinkText",
  \ "markdownLinkTextDelimiter",
  \ "mkdLink",
  \ "mkdDelimiter"
  \ ]

...

function! OnMarkdownLink()
  let current_syntax_id = synIDattr(synID(line("."), col("."), 1), "name")
  return count(s:MARKDOWN_LINK_SYNTAX_IDS, current_syntax_id) != 0
endfunction
```

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+endf
+
+function! s:OpenMarkdownLink()
+  "get the initial position of the curosr
```
I think the next line is clear enough without this comment.

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+  "empty the @a register in the case cursor is not well palced and that old
+  "patterns would still trigger.
+  let @a = ""
+  normal "ayi]
```
Extract these to a function `s:CaptureLinkText` to clean up this function?

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+  let found = search(link_target_pattern, 'e')
+  if found
+    "use netrw-gx to open the link
+    call netrw#BrowseX(expand("<cfile>"),0)
```
Extract to a function link `s:OpenWithNetrw()` since we reuse?

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+  let escaped_link_name = escape(getreg('a'), '&')
+  let link_target_pattern = '\v^\['.escaped_link_name.'\]: (%(ftp[s]?|http[s]?):\/\/\S+)>'
+  " go to link location
+  let found = search(link_target_pattern, 'e')
```
I think this is clearer is you inline the search call into the `if`, so it would just be:

``` vim
if search(link_target_pattern, 'e')
```

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +95,45 @@ function! OpenLinkOnCurrentLine()
+endif
+if g:quicklink_open_mapping != ''
+  execute 'nnoremap <buffer> '.g:quicklink_open_mapping.' :MarkdownAwareGX<cr>'
+endif
```
Now that we are only augmenting `gx` rather than providing a distinct mapping, I'd prefer to just use that. If user's want a custom map they can use the command you've defined on 131. Then these lines can just be:

``` vim
command! MarkdownAwareGX call <sid>MarkdownAwareGX()
nnoremap <buffer> gx :MarkdownAwareGX<cr>
```

christoomey commented on ftplugin/markdown/quicklink.vim:
```diff
@@ -95,4 +107,40 @@ function! OpenLinkOnCurrentLine()
+  if s:OnMarkdownLink()
+    call s:OpenMarkdownLink()
+  else
+    call netrw#BrowseX(expand("<cfile>"),0)
```
I'm getting an error running this locally. I think it needs to be `call netrw#NetrwBrowseX(expand("<cfile>"), 0)`

sim590 replied:
We can patch this using this : 
```vim
if has("patch-7.4.567")
  function! s:Browse(args)
    execute 'call netrw#BrowseX('.string(a:args).', 0)'
  endfunction

  command! -nargs=* Browse call <SID>Browse(<q-args>)
endif
```
I took it from the post I linked.

christoomey replied:
That sounds good. Looks like you've done that in the `s:OpenWithNetrw` function above, but you don't seem to be calling that function here. Am I missing something?

sim590 replied:
I've wrote it too quickly. I've fixed it in c0880dd.

sim590 commented:
It was giving me an error when putting `netrw#NetrwBrowseX`, so I've looked at the code in `$VIMRUNTIME/autoload/netrw.vim` : http://ix.io/gox. But you're not the first that tells me that `netrw#NetrwBrowseX` works... I don't know what's going on... I'll look it up.

sim590 commented:
What's your vim version ? Mine's 7.4.617. It appears that the function has been renamed in a patch :
https://github.com/tpope/vim-fugitive/issues/594

sim590 commented:
I've added support for opening the link when the cursor is positionned like this :

christoomey commented:
Nice trick to support the link definition! I think you need to add support for the additional syntaxIDs for link definitions. On my setup it is "markdownIdDeclaration", but I notice that you are also adding "mkd" prefixed versions so I am guessing there is a related "mkd" syntax ID you'll want to add.

Thanks for putting up with all my edits and suggestions by the way!

sim590 commented:
I've added link definitions in e76deee and I also added a command in 356d415 for jumping to the link definition which I found was useful for jumping at the end of the section where all the link definitions are placed.

christoomey commented:
Squashed and merged as https://github.com/christoomey/vim-quicklink/commit/361b1ef91272f66349988b8c893f2eb6d0e2c96f. Awesome work @sim590! Thanks for taking this on.

sim590 commented:
That was a pleasure! It gave me the opportunity to get into vim plugin writing, which I had not done really. Thank you for accepting my changes!

christoomey commented:
My pleasure as well, and you certainly took to writing vimscript quite well for your first time. Thanks again!
